TechnicalDocuments/Proposals/StyleSheets.md

```diff
@@ -0,0 +1,119 @@
+# Composable value-type style sheets
+
+* Author(s): Anders Ha
+* Review Manager: Sergey Shulga, David Rodrigues
+
+## Problem Statement
+The current inheritance based StyleSheet approach is unideal in three areas:
+
+1. Ease of Use: Requiring upfront declarations of all styleable properties.
+
+    We would have to constantly match UIKit changes. Custom views would require manual implementation.
+
+2. Mental Model: Awkwardness when being applied to views that are opened for extension.
+
+    When one defines a non generic `ExtensibleViewStyleSheet` for an `ExtensibleView`, the `StyleSheetProtocol.Element` associated type is immediately satisfied with the concrete view type. This means in a generic environment bound to `StyleSheetProtocol`, only `apply(to _: ExtensibleView)` would be recognised since `Element == ExtensibleView` per the conformance. In other words, all stylesheet subclasses must override this base class typed method & are required to downcast at runtime to access stuff defined in the subclass.
+
+    On the other hand, when one defines a generic `ExtensibleViewStyleSheet<T: ExtensibleView>` stylesheet, it does address the said issue by satisfying `StyleSheetProtocol.Element` with a generic parameter. However, it leads to awkward and verbose spellings at use sites in general, e.g. `ExtensibleViewStyleSheet<ExtensibleView>`.
+
+3. Correctness: Unintended Sharing
+
+    Stylesheets are currently classes so as to be able to take advantage of inheritance, and this makes them prone to unintended sharing. With Swift offering first-class value type, this might amplify its possibility due to common perception of configurations being modelled as value types.
+
+## Identified Requirements
+
+1. The approach must be applicable to reusable views.
+
+    When multiple style sheets with symmetrical differences are applied in a row in any order, the reused view would not end up in an indeterministic state.
+
+1. The approach must be compatible with view subclassing.
+
+1. The approach should address unintended sharing if possible.
+
+## Proposed Solution
+
+A new value type `StyleSheet<View>` shall be introduced, with the following features:
+
+1. Copy on write & value semantics: Backed by CoW `Dictionary`. Support `Equatable` out of the box.
+
+2. Built around Key Paths introduced since Swift 4, and has proper support of entries setting key paths that are partially overlapping with each other.
+
+3. Support producing **inverses** for deterministic view reuse.
```

Comment:
I'm not a fan of the term "inverse" in that context, as in my mind that carries the idea that the function inverse in the mathematical sense is computed in some magic way. Like magically creating a `view.alpha *= 2.0` function from a `view.alpha *= 0.5` function.

I think the term "snapshot" that you used in the description is more accurate and understandable —and less confusing— to describe some state that will be _restored_ / _restorable_ before the next call to `apply` or during `prepareForReuse()`